src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1096,6 +1095,30 @@ public virtual IHtmlContent GenerateAntiforgery(ViewContext viewContext)
             return currentValues;
         }
 
+        // Public to minimize churn when (2.0?) this is added to IHtmlGenerator.
+        /// <summary>
+        /// Returns a valid HTML 4.01 "id" attribute value for an element with the given <paramref name="fullName"/>.
+        /// </summary>
+        /// <param name="viewContext">A <see cref="ViewContext"/> instance for the current scope.</param>
+        /// <param name="fullName">
+        /// The fully-qualified expression name, ignoring the current model. Also the original HTML element name.
+        /// </param>
+        /// <returns>
+        /// Valid HTML 4.01 "id" attribute value for an element with the given <paramref name="fullName"/>.
+        /// </returns>
+        /// <seealso cref="TagBuilder.CreateSanitizedId"/>
+        public virtual string CreateSanitizedId(ViewContext viewContext, string fullName)
```

Comment:
`IHtmlGenerator.GenerateGroupsAndOptions()` is the only method on this API that does _not_ have a `ViewContext` first parameter.

One option for both `GenerateId()` and `CreateSanitizedId()` would be to `[Obsolete]` both in `TagBuilder` and duplicate the low-level bits of `CreateSanitzedId()` here.

Could also avoid `HtmlHelper` downcasts w/ a new `interface`. Implementation could be in a new service or in `DefaultHtmlGenerator`. In the second case, might add an `HtmlGeneratorAdapter` to cover the case where user replaces the `IHtmlGenerator` implementation in DI.